xla/service/gpu/gpu_compiler.cc

```diff
@@ -2210,22 +2210,44 @@ absl::Status GpuCompiler::RunPostSchedulingPipelines(
   {
     HloPassPipeline pipeline("remat-pipeline");
 
-    HloCostAnalysis hlo_cost_analysis(ShapeSizeBytesFunction());
+    int64_t offloading_memory_limit_gb =
+        module->config().debug_options().xla_gpu_offloading_memory_limit_gb() *
+        1024 * 1024 * 1024;
+    float flops_per_sec = gpu_device_info.core_count() *
+                          gpu_device_info.fpus_per_core() *
+                          gpu_device_info.clock_rate_ghz() * /*giga*/1e+9 * 2;
+    int64_t host_memory_space_color =
+        static_cast<int64_t>(se::MemoryType::kHost);
+    bool enable_offloading = offloading_memory_limit_gb > 0;
+
     HloRematerialization::RematerializationModeConfig
         rematerialization_mode_config(/*recompute=*/true, /*compress=*/true,
-                                      /*host_offload=*/false);
+                                      /*host_offload=*/enable_offloading);
+    HloCostAnalysis::Options hlo_cost_analysis_options;
+    hlo_cost_analysis_options.shape_size = ShapeSizeBytesFunction();
+    if (enable_offloading) {
+      hlo_cost_analysis_options.set_flops_per_second(flops_per_sec);
+      hlo_cost_analysis_options.set_transcendentals_per_second(flops_per_sec);
+    }
+    HloCostAnalysis hlo_cost_analysis(hlo_cost_analysis_options);
+    HloRematerialization::HostMemoryOffloadConfig offloading_config(
```

Comment:
If offloading is not enabled, you should not be constructing a host memory offloading config nor passing it to the HloRematerialization options.

I think it is completely ok if you do, but it's an absolute waste and maybe confusing to someone else.

Reply:
> If offloading is not enabled, you should not be constructing a host memory offloading config nor passing it to the HloRematerialization options.
How to pass a std::nullopt and offloading_config in an implicit if statement to construct a remat pass? I couldn't solve the type mismatch. It'll be nice to do it.

Reply:
```
    HloCostAnalysis::Options hlo_cost_analysis_options;
    hlo_cost_analysis_options.shape_size = ShapeSizeBytesFunction();
    std::optional<HloRematerialization::HostMemoryOffloadConfig> offloading_config;
    if (enable_offloading) {
      hlo_cost_analysis_options.set_flops_per_second(flops_per_sec);
      hlo_cost_analysis_options.set_transcendentals_per_second(flops_per_sec);
      offloading_config.emplace(/*host_memory_space=*/host_memory_space_color,
                                /*bandwidth_to_host_bytes_per_second=*/
                                gpu_device_info.memory_bandwidth(),
                                /*bandwidth_from_host_bytes_per_second=*/
                                gpu_device_info.memory_bandwidth());
    }
    HloCostAnalysis hlo_cost_analysis(hlo_cost_analysis_options);
    ```
    Then pass offloading_config unconditionally to the options struct

Reply:
Thanks so much for showing me how to write such a code. I made a similar version and verified.